overlord/exportstate/manifest.go

```diff
@@ -0,0 +1,211 @@
+	"path/filepath"
+
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+// Manifest describes content content exported to snaps or the host.
+type Manifest struct {
+	SnapName        string          `json:"snap-name"`
+	ExportedVersion string          `json:"exported-version"`
+	Symlinks        []SymlinkExport `json:"symlinks,omitempty"`
+}
+
+// NewManifestForSnap returns the manifest of a given snap.
+//
+// Currently only the core and snapd snaps export any content to the system. As
+// such the export manifest is not embedded into the snap meta-data but instead
+// computed here.
+//
+// Both snapd and core snaps export content under the snap name "snapd", using
+// the export set name "tools". The revision is mangled, for "snapd" it is used
+// directly. For "core" it is transformed to "core_$revision".
+func NewManifestForSnap(info *snap.Info) *Manifest {
+	// XXX: should we use WellKnownSnapID here? Probably not as this must work for
+	// unsigned snaps as well. Alternatively, should we look at snap type, as we
+	// have unique values for both snapd and core.
+	switch info.SnapName() {
+	case "snapd":
+		return manifestForSnapdSnap(info)
+	case "core":
+		return manifestForCoreSnap(info)
+		// XXX: do we need to handle ubuntu-core?
+	default:
+		return manifestForRegularSnap(info)
+	}
+}
+
+// NewManifestForHost returns the export manifest of the host system.
+//
+// Ubuntu Core systems do not have an export manifest. Classic systems have an export
+// manifest which describes snapd tools from the classic package.
+func NewManifestForHost() *Manifest {
+	if release.OnClassic {
+		return manifestForClassicSystem()
+	}
+	return manifestForCoreSystem()
+}
+
+// IsEmpty returns true if a manifest describes no content.
+func (m *Manifest) IsEmpty() bool {
+	return len(m.Symlinks) == 0
+}
+
+// CreateExportedFiles creates all the files constituting the export manifest.
+//
+// The directory /var/lib/snapd/export/$snapName/$exportedVersion is created
+// if necessary. For each export set in the manifest, additional sub-directory
+// is created and populated with symbolic links pointing to the exported files.
+//
+// The function is idempotent.
+func (m *Manifest) CreateExportedFiles() error {
+	for _, s := range m.Symlinks {
+		if err := s.Create(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RemoveExportedFiles removes all the files constituting the export state.
+//
+// In addition the path /var/lib/snapd/export/$snapName/$exportedVersion
+// is pruned, removing empty directories if possible.
+//
+// On failure removal continues and the first error is returned.
+func (m *Manifest) RemoveExportedFiles() error {
+	var firstErr error
+	for _, s := range m.Symlinks {
+		if err := s.Remove(); err != nil && firstErr == nil {
+			firstErr = err
+		}
+	}
+	return firstErr
+}
+
+// SymlinkExport describes content exported as symbolic link.
+type SymlinkExport struct {
+	SnapName        string `json:"snap-name"`
```

Comment:
there is some convenience but also messiness related to repeating the SnapName and ExportedVersion here. My preference would be to omit these,  and instead pass a Manifest or the values to the methods below.

Reply:
Sure, I see the point. That's doable for sure.

Reply:
This is addressed by the new Manifest type set.